### ISSUE TITLE
Protocol Error

### ISSUE THREAD

**Issue by ankhers** [elixir-mongo/mongodb#75]
This is for the emj-dbconn branch

I believe my recent PR (#74) broke something. When trying to do simple queries, I am getting a protocol error.

However, if I execute the commands fast enough (copy / paste into `iex`) the query works properly. 

Let me know if there is anything else I can do to debug this.

```
12:44:41 (emj-dbconn) mongodb$ iex -S mix
Erlang/OTP 19 [erts-8.0] [source] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false]

Interactive Elixir (1.3.0) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> {:ok, pid} = Mongo.start_link(database: "some_database")
{:ok, #PID<0.154.0>}
iex(2)> c = Mongo.find(pid, "users", %{})
%Mongo.Cursor{coll: "users", conn: #PID<0.154.0>, opts: [], query: %{},
 select: nil}
iex(3)> Enum.to_list(c)
** (Mongo.Error) tcp recv: invalid argument - :einval

12:45:27.973 [info]  Mongo.Protocol (#PID<0.154.0>) missed message: {:tcp, #Port<0.5043>, <<167, 25, 7, 0, 1, 245, 4, 0, 0, 0, 0, 0, 1, 0, 0, 0, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 143, 1, 0, 0, 50, 24, 0, 0, 7, 95, 105, 100, 0, 82, 129, 95, ...>>}

12:45:27.973 [error] Mongo.Protocol (#PID<0.154.0>) disconnected: ** (Mongo.Error) tcp recv: invalid argument - :einval
    (mongodb) lib/mongo/cursor.ex:40: anonymous fn/6 in Enumerable.Mongo.Cursor.start_fun/6
     (elixir) lib/stream.ex:1121: anonymous fn/5 in Stream.resource/3
     (elixir) lib/enum.ex:1626: Enum.reduce/3
     (elixir) lib/enum.ex:2345: Enum.to_list/1
** (EXIT from #PID<0.152.0>) an exception was raised:
    ** (RuntimeError) disconnect/2 not implemented
        (mongodb) lib/db_connection.ex:240: Mongo.Protocol.disconnect/2
        (db_connection) lib/db_connection/connection.ex:178: DBConnection.Connection.disconnect/2
        (connection) lib/connection.ex:767: Connection.disconnect/3
        (stdlib) gen_server.erl:601: :gen_server.try_dispatch/4
        (stdlib) gen_server.erl:667: :gen_server.handle_msg/5
        (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3

Interactive Elixir (1.3.0) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)>
12:45:27.978 [error] GenServer #PID<0.154.0> terminating
** (RuntimeError) disconnect/2 not implemented
    (mongodb) lib/db_connection.ex:240: Mongo.Protocol.disconnect/2
    (db_connection) lib/db_connection/connection.ex:178: DBConnection.Connection.disconnect/2
    (connection) lib/connection.ex:767: Connection.disconnect/3
    (stdlib) gen_server.erl:601: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:667: :gen_server.handle_msg/5
    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
Last message: {:"$gen_cast", {:disconnect, #Reference<0.0.6.118>, %Mongo.Error{code: nil, message: "tcp recv: invalid argument - :einval"}, %{database: "some_database", request_id: 0, socket: #Port<0.5043>, timeout: 5000, wire_version: 4, write_concern: %{w: 1}}}}
State: {Mongo.Protocol, %{database: "some_database", request_id: 0, socket: #Port<0.5043>, timeout: 5000, wire_version: 4, write_concern: %{w: 1}}}

```

**Comment by ankhers** [comment]
Just some additional information after speaking with fishcakez on slack. It appears that this is an issue with passive / active on the socket.

> ok
> https://gist.github.com/ankhers/b08420a94575c7eb2f26c23e61380e0d#file-gistfile1-txt-L13
> here you have "messed up" active/passive on the socket, should be passive but are still active and a message arrived in the connection process while the client process had control of socket
> you get the tcp error :einval because you are trying to receive on an active socket with a passive :gen_tcp.recv
> then you crash because disconnect/2 is not implemented
> 
> https://github.com/ericmj/mongodb/blob/e84c1b291550ddb3c407a3c80061da25322175d3/lib/mongo/protocol.ex#L101 there is a bug here
> see https://github.com/elixir-ecto/postgrex/blob/master/lib/postgrex/protocol.ex#L117
> you need to flush the mesage queue after setting active false

**Comment by ankhers** [comment]
This was fixed with #77
